pyproject.toml

```diff
@@ -307,6 +307,11 @@ extend-exclude = [
     "tests/dataset",                                            # 10 warnings
 ]
 
+[tool.ruff.lint]
+# This enables preview rules for specified rules e.g. NPY201
+preview = true
+explicit-preview-rules = true  # https://docs.astral.sh/ruff/preview/#selecting-single-preview-rules
```

Comment:
This is where we enable preview rules. Besides the per file ignore below, the rest of the changes in this PR are issues found by enabling these preview rules (or the preview version of existing rules).